runtime/list.go

```diff
@@ -291,6 +291,37 @@ func listNE(f *Frame, v, w *Object) (*Object, *BaseException) {
 	return listCompare(f, toListUnsafe(v), w, NE)
 }
 
+func listPop(f *Frame, args Args, kwargs KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	expectedTypes := []*Type{ListType, ObjectType}
+	if argc == 1 {
+		expectedTypes = expectedTypes[:1]
+	}
+	if raised := checkMethodArgs(f, "pop", args, expectedTypes...); raised != nil {
+		return nil, raised
+	}
+	i := -1
+	if argc == 2 {
+		o, raised := IntType.Call(f, Args{args[1]}, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		i = toIntUnsafe(o).Value()
+	}
+	l := toListUnsafe(args[0])
+	if i < 0 {
+		i += len(l.elems)
+	}
+	if i >= len(l.elems) || i < 0 {
+		return nil, f.RaiseType(IndexErrorType, "list index out of range")
+	}
+	l.mutex.Lock()
```

Comment:
Thanks for the explanations and implementation. Much better understanding on mutex now.